flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java

```diff
@@ -250,15 +250,20 @@ public void release(Throwable cause) {
     /** Releases all produced data including both those stored in memory and persisted on disk. */
     protected abstract void releaseInternal();
 
-    @Override
-    public void close() {
+    private void closeBufferPool() {
         if (bufferPool != null) {
             bufferPool.lazyDestroy();
         }
     }
 
+    @Override
+    public void close() {
+        closeBufferPool();
+    }
+
     @Override
     public void fail(@Nullable Throwable throwable) {
+        closeBufferPool();
```

Comment:
I don't fully understand the contract here. When ResultParitition fails should we close it or not? According to your code, we only close the buffer pool but don't close the result partition in general. I am afraid that it can lead to another leak if tomorrow we will add something to `close` then `fail` method closes the buffer pool but won't do another things related to `close`. 
Perhaps, it is not the discussion for now but maybe it is better to rename `fail` to `close` then we will have more clear contract - `close` and `close(Throwable)`

Reply:
In my understanding the close method of result partition will be always called by the task thread to release the network resources regardless of success or failure. I think for network resources releasing, it is enough. I guess the only reason to call the close method in the TaskCanceler thread is as the comment suggested:
>>>                // Early release of input and output buffer pools. We do this
>>>                // in order to unblock async Threads, which produce/consume the
>>>                // intermediate streams outside of the main Task Thread (like
>>>                // the Kafka consumer).
I did not look into this and just moved the BufferPool close operation to the fail method to avoid breaking this behavior.

About the fail method, I think it more likes another release method without knowing the partition id.